contracts/hydro/src/contract.rs

```diff
@@ -1624,6 +1651,99 @@ fn compute_round_end(constants: &Constants, round_id: u64) -> StdResult<Timestam
+    info: &MessageInfo,
+    constants: &Constants,
+    current_round: u64,
+    old_lock_entry: Option<LockEntry>,
+    new_lock_entry: LockEntry,
+    validator: String,
+) -> Result<(), ContractError> {
+    let round_end = compute_round_end(constants, current_round)?;
+    let lock_epoch_length = constants.lock_epoch_length;
+
+    let old_scaled_shares = match old_lock_entry {
+        None => Uint128::zero(),
+        Some(lock_entry) => {
+            get_lock_time_weighted_shares(round_end, lock_entry.clone(), lock_epoch_length)
+        }
+    };
+    let new_scaled_shares =
+        get_lock_time_weighted_shares(round_end, new_lock_entry, lock_epoch_length);
+
+    let vote_power_change = Decimal::from_ratio(new_scaled_shares, Uint128::one())
+        .checked_sub(Decimal::from_ratio(old_scaled_shares, Uint128::one()))?;
+
+    let tranche_ids = TRANCHE_MAP
+        .keys(deps.storage, None, None, Order::Ascending)
+        .collect::<Result<Vec<u64>, StdError>>()?;
+
+    for tranche_id in tranche_ids {
+        if let Some(mut vote) = VOTE_MAP.may_load(
+            deps.storage,
+            (current_round, tranche_id, info.sender.clone()),
+        )? {
+            let current_vote_shares = match vote.time_weighted_shares.get(&validator) {
+                None => Decimal::zero(),
+                Some(shares) => *shares,
+            };
+
+            let new_vote_shares = current_vote_shares.checked_add(vote_power_change)?;
+            vote.time_weighted_shares
+                .insert(validator.clone(), new_vote_shares);
+
+            VOTE_MAP.save(
+                deps.storage,
+                (current_round, tranche_id, info.sender.clone()),
+                &vote,
+            )?;
+
+            add_validator_shares_to_proposal(
+                deps.storage,
+                current_round,
+                vote.prop_id,
+                validator.clone(),
+                vote_power_change,
+            )?;
+
+            let proposal_id = vote.prop_id;
+            let mut proposal =
+                PROPOSAL_MAP.load(deps.storage, (current_round, tranche_id, proposal_id))?;
+
+            PROPS_BY_SCORE.remove(
+                deps.storage,
+                (
+                    (current_round, tranche_id),
+                    proposal.power.into(),
+                    proposal_id,
+                ),
+            );
+
+            let total_power = get_total_power_for_proposal(deps.storage, proposal_id)?;
+            proposal.power = total_power.to_uint_ceil();
+
+            PROPOSAL_MAP.save(
+                deps.storage,
+                (current_round, tranche_id, proposal_id),
+                &proposal,
+            )?;
+
+            PROPS_BY_SCORE.save(
+                deps.storage,
+                (
+                    (current_round, tranche_id),
+                    proposal.power.into(),
+                    proposal_id,
+                ),
+                &proposal_id,
+            )?;
```

Comment:
I extracted everything except `add_validator_shares_to_proposal()` because this part is different in `vote()` and `update_voting_power_on_proposals()` (one has iteration, the other has condition and a call to  `remove_validator_shares_from_proposal()`, so I didn't want to make the new utility function too complex).